File: String/reverseVowels/reverseVowels.cpp
#include <iostream>
#include <string>

std::string reverse_vowels(std::string str) {
	if (str.size() == 0) return str; // Edge case since we are using size_type (unsigned)
	
	typedef std::string::size_type str_size;
	str_size i = 0;
	str_size j = str.size() - 1;
	
	while (i < j) {
		i = str.find_first_of("aeiouAEIOU", i);
		j = str.find_last_of("aeiouAEIOU", j);
		// Make sure we swap from left to right only
		if (i < j) {
			char temp = str[i];
			str[i] = str[j];
			str[j] = temp;
			++i; ++j;
		}
	}
	
	return str;
}

int main() {
  std::cout << "Reverse vowels 'Hello': " << reverse_vowels("Hello") << std::endl;
  std::cout << "Reverse vowels 'VOwEls': " << reverse_vowels("VOwEls") << std::endl;
  std::cout << "Reverse vowels: 'EECS 281': " << reverse_vowels("EECS 281") << std::endl;
}

File: String/reverseVowels/reverseVowels.go
package main

import "fmt"

func reverseVowels(s string) string {
	if len(s) <= 1 {
		return s
	}
	runes := []rune(s)
	front, back := 0, len(runes)-1
	for front < back {
		for !isVowel(runes[front]) {
			front++
		}
		for !isVowel(runes[back]) {
			back--
		}
		if front < back {
			temp := runes[front]
			runes[front] = runes[back]
			runes[back] = temp
			front++
			back--
		}
	}
	return string(runes)
}

func isVowel(r rune) bool {
	switch r {
	case 'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U':
		return true
	}
	return false
}

func main() {
	reverseVowelsPrint("Hello")    // Holle
	reverseVowelsPrint("VOwEls")   // VEwOls
	reverseVowelsPrint("EECS 281") // EECS 281
}

func reverseVowelsPrint(s string) {
	fmt.Printf("Reverse vowels %s: %s\n", s, reverseVowels(s))
}
